src/Processors/Formats/Impl/ParallelFormattingOutputFormat.h

```diff
@@ -0,0 +1,265 @@
+            collector_thread.join();
+
+        try
+        {
+            pool.wait();
+        }
+        catch (...)
+        {
+            tryLogCurrentException(__PRETTY_FUNCTION__);
+        }
+    }
+
+
+    void onBackgroundException()
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+
+        std::unique_lock<std::mutex> lock(mutex);
+        if (!background_exception)
+        {
+            background_exception = std::current_exception();
+        }
+        formatting_finished = true;
+        writer_condvar.notify_all();
+        collector_condvar.notify_all();
+    }
+
+    void scheduleFormatterThreadForUnitWithNumber(size_t ticket_number)
+    {
+        pool.scheduleOrThrowOnError([this, ticket_number] { formatterThreadFunction(ticket_number); });
+    }
+
+    void waitForFormattingFinished()
+    {
+        ///FIXME
+        while(hasChunksToWorkWith())
+        {
+            std::cout << writer_unit_number << ' ' << collector_unit_number << std::endl;
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        }
+    }
+
+    bool hasChunksToWorkWith()
+    {
+        return writer_unit_number - collector_unit_number > 0;
+    }
+
+    void collectorThreadFunction()
+    {
+        setThreadName("Collector");
+
+        try
+        {
+            while (!formatting_finished)
+            {
+                const auto current_unit_number = collector_unit_number % processing_units.size();
+
+                auto &unit = processing_units[current_unit_number];
+
+                {
+                    std::unique_lock<std::mutex> lock(mutex);
+                    collector_condvar.wait(lock,
+                        [&]{ return unit.status == READY_TO_READ || formatting_finished; });
+                }
+
+                if (formatting_finished)
+                    break;
+
+                assert(unit.status == READY_TO_READ);
+                assert(unit.segment.size() > 0);
+
+                /// Do main work here.
+                out.write(unit.segment.data(), unit.actual_memory_size);
+                out.sync();
+
+                ++collector_unit_number;
+
+                std::lock_guard<std::mutex> lock(mutex);
+                unit.status = READY_TO_INSERT;
+                writer_condvar.notify_all();
+            }
+        }
+        catch (...)
+        {
+            onBackgroundException();
+        }
+    }
+
+
+    void formatterThreadFunction(size_t current_unit_number)
+    {
+        setThreadName("Formatter");
+
+        try
+        {
+            auto & unit = processing_units[current_unit_number];
+
+            assert(unit.status = READY_TO_FORMAT);
+
+            unit.segment.resize(DBMS_DEFAULT_BUFFER_SIZE);
+
+            /// TODO: Implement proper nextImpl
+            BufferWithOutsideMemory<WriteBuffer> out_buffer(unit.segment);
+
+            auto formatter = internal_formatter_creator(out_buffer);
+
+            formatter->consume(std::move(unit.chunk));
```

Comment:
It may not work correctly for formats with non-trivial `doWritePrefix()` and `doWriteSuffix()`. Maybe disable parallel formatting for such formats?

Reply:
For now I've added `doWritePrefix()` and `doWriteSuffix()` methods, but it won't still work correctly for formats with stateful parsers (such as JSON). But it has to work with `CSVWithNamesAndTypes`, etc.